core/src/main/java/org/apache/calcite/rel/rules/FilterWindowTransposeRule.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Window;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import com.google.common.collect.ImmutableList;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that pushes a {@link org.apache.calcite.rel.core.Filter}
+ * past a {@link org.apache.calcite.rel.core.Window}.
+ *
+ * <p> If {@code Filter} condition used columns belongs {@code Window} partition keys,
+ * then we could push the condition past the {@code Window}.
+ *
+ * <p> For example:
+ *  <blockquote><pre>{@code
+ *    LogicalProject(NAME=[$0], DEPTNO=[$1], EXPR$2=[$2])
+ *     LogicalFilter(condition=[>($1, 0)])
+ *       LogicalProject(NAME=[$1], DEPTNO=[$0], EXPR$2=[$2])
+ *         LogicalWindow(window#0=[window(partition {0} aggs [COUNT()])])
+ *           LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+ *  }</pre></blockquote>
+ *
+ * <p> will convert to:
+ * <blockquote><pre>{@code
+ *    LogicalProject(NAME=[$1], DEPTNO=[$0], EXPR$2=[$2])
+ *      LogicalWindow(window#0=[window(partition {0} aggs [COUNT()])])
+ *        LogicalFilter(condition=[>($0, 0)])
+ *          LogicalTableScan(table=[[CATALOG, SALES, DEPT]])
+ * }</pre></blockquote>
+ *
+ * @see CoreRules#FILTER_PROJECT_TRANSPOSE
+ * @see CoreRules#PROJECT_TO_LOGICAL_PROJECT_AND_WINDOW
+ * @see CoreRules#FILTER_WINDOW_TRANSPOSE
+ */
+@Value.Enclosing
+public class FilterWindowTransposeRule
```

Comment:
The rule resembles a lot the `FilterAggregateTransposeRule`. Would it be possible to re-use some of the logic by refactoring the rule(s)? Maybe keeping one rule and varying the configurations?

Reply:
Yes,It make sense for me.

In terms of the use of rules, I will still divide them into two Rules, but I will reuse the similar logic of the two rules as much as possible.

My idea is to put similar logic into RelOptUtil.

Reply:
Thinking about it again, I would suggest to postpone the refactoring in a dedicated follow-up ticket.This would allow this PR to be merged faster; refactoring can more debatable so let's evaluate it separately :)

Reply:
Get your point,I will take the advice.